modernize ControllerManager

Swiftb0y opened the pull request:
And a little bit of the surrounding code. This is what I have so far, though the multithreading code could use a little larger restructuring. based on (outdated) #13622.

Swiftb0y committed:
refactor: modernize bulksupport allowlist

* move equality comparision closer to type definition.
* remove trailing null-entry in support array
* modernize allowlist search

Swiftb0y committed:
refactor: modernize hid_denylist

* replace magic (unnamed) values with proper constants
* use idiomatic array iteration

Swiftb0y committed:
refactor: return string_views instead of char ptrs for hiddevice

Swiftb0y committed:
refactor: improve HidEnumerator memory safety

Swiftb0y committed:
refactor: improve memory-safety of BulkEnumerator

Swiftb0y committed:
refactor: cleanup MidiEnumerator

Swiftb0y committed:
refactor: cleanup ControllerEnumerator

Swiftb0y committed:
refactor: improve controllerCompare safety

Swiftb0y committed:
refactor: use `<chrono>` instead of `mixxx::Duration`

Swiftb0y committed:
refactor: improve ControllerManager memory safety

* explicitly initialize all members
* use unique_ptr for owning pointers
* use std::algorithm where applicable
* use compile-time literals where possible

Swiftb0y committed:
refactor: replace bare QDebug() with `mixxx::logger`

Swiftb0y committed:
Revert "refactor: return string_views instead of char ptrs for hiddevice"

Because qt is too old...

This reverts commit b021f641c49bed472ff5c5381e6e1b5f79747121.

Swiftb0y committed:
fixup! refactor: improve memory-safety of BulkEnumerator

Swiftb0y committed:
refactor: replace owning QThread pointer with unique_ptr

Swiftb0y committed:
refactor: inline trivial getters and improve const-correctness

Swiftb0y committed:
refactor: mark privately-used member functions private

Swiftb0y committed:
refactor: document threading assumptions using thread-affinity asserts

Swiftb0y commented on src/controllers/hid/hidenumerator.cpp:
```diff
@@ -120,10 +118,13 @@ QList<Controller*> HidEnumerator::queryDevices() {
 
-        HidController* newDevice = new HidController(std::move(deviceInfo));
-        m_devices.push_back(newDevice);
+        m_devices.push_back(std::make_unique<HidController>(std::move(deviceInfo)));
```
not that independent of my changes, this is currently broken when querying devices multiple times because we don't clear the list from the previous query... Not sure if I should fix that here because its not feasible to test nor do I have enough non-keyboard/-mice HID devices here.

daschuer replied:
I think we should fix it anyway. Main is still in alpha stage.

daschuer reviewed (changes requested):
some comments of #13627 apply here as well. Can you adopt them?

Swiftb0y commented:
> some comments of https://github.com/mixxxdj/mixxx/pull/13627 apply here as well. Can you adopt them?

I assume you mean #13622? Yes, I will rebase on that once everything is happy.

acolombier reviewed (commented):
Looks good overall! One request tho, could you make sure that updated pointer variables are prefixed with `p...` to help with readability? 

One mild concerned is around BULK, not directly related to your change: Interface also appears to be relevant on Linux, and not using them leads to `dmesg` warnings, so I'm wondering if would be worth removing the preprocessor around this. If you're keen to give it a try, I'm happy to test it with my devices to confirm it works as expected!

Swiftb0y commented:
Will look into both. Thanks for taking a look @acolombier

Swiftb0y committed:
fixup! refactor: improve ControllerManager memory safety

Swiftb0y committed:
fixup! refactor: improve memory-safety of BulkEnumerator

Swiftb0y committed:
fixup! refactor: improve HidEnumerator memory safety

Swiftb0y committed:
fixup! refactor: improve ControllerManager memory safety

Swiftb0y committed:
refactor: automate memory management in PortMidiEnumerator

Swiftb0y committed:
refactor: optimize PortMidi namematching by delaying QString creation

Swiftb0y committed:
document that queryDevices will invalidate previous device lists

Swiftb0y committed:
clear device list on queryDevices for HID and Bulk enumerators

This fixes a (hypothetical) issue which could cause duplicate
HID and Bulk devices to be created when querying for devices
multiple times. This has not not happened yet because devices
so far have only been queried once on startup.

Swiftb0y committed:
refactor: modernize Hss1394Enumerator

Swiftb0y committed:
refactor: remove unused UserSettingsConfig from BulkEnumerator

Swiftb0y committed:
fixup! refactor: optimize PortMidi namematching by delaying QString creation

acolombier commented on src/controllers/midi/portmidienumerator.cpp:
```diff
@@ -1,13 +1,15 @@
 #include "controllers/midi/portmidienumerator.h"
 
 #include <portmidi.h>
+#include <qlatin1stringview.h>
```
I belieb you should use `#include <QLatin1String>` instead here

acolombier commented on src/controllers/midi/portmidienumerator.cpp:
```diff
@@ -1,13 +1,15 @@
 #include "controllers/midi/portmidienumerator.h"
 
 #include <portmidi.h>
+#include <qlatin1stringview.h>
```
What is this used for? Looks like it doesn't exist on MacOS

Swiftb0y replied:
whoops, yeah clangd tries its best to add the relevant includes but its gets it wrong and I don't notice it. I'll remove and force push (and look into disabling that for certain headers).

Swiftb0y replied:
done

Swiftb0y replied:
fyi in case any1 else stumbles over this, see https://clangd.llvm.org/guides/include-cleaner#adjusting-behavior and use `IgnoreHeaders: q.*.h` until [QTBUG-124159](https://bugreports.qt.io/browse/QTBUG-124159) is fixed.

daschuer commented:
Some conflicts have developed.

daschuer commented on src/controllers/bulk/bulkenumerator.cpp:
```diff
@@ -2,60 +2,73 @@
-        }
-    }
-    return false;
+static bool is_interesting(const libusb_device_descriptor& desc) {
```
This can go to an anonymous namespace

daschuer commented on src/controllers/bulk/bulkenumerator.cpp:
```diff
@@ -2,60 +2,73 @@
+    auto pDeviceList = std::unique_ptr<libusb_device*[],
+            decltype([](libusb_device** list) {
+                libusb_free_device_list(list, true);
+            })>(deviceListPointers);
```
This looks bulky. How about creating `deviceListDeleter()` in the anonymous namespace?

daschuer commented on src/controllers/bulk/bulkenumerator.cpp:
```diff
@@ -2,60 +2,73 @@
+        // of the arguments in the case of errors.
+        qWarning() << "Error retrieving devicelist: " << libusb_error_name(cnt);
+    }
+    auto pDeviceList = std::unique_ptr<libusb_device*[],
```
Wy using only here `libusb_device*[]` and not above?

daschuer commented on src/controllers/bulk/bulkenumerator.cpp:
```diff
@@ -2,60 +2,73 @@
+    }
+    QList<Controller*> devices;
+    devices.reserve(m_devices.size());
+    for (const auto& pController : std::as_const(m_devices)) {
```
Can we get rid of this loop by returning a const refrences to m_devices? Before it was not an issue because of QTs implicit sharing.

Swiftb0y replied:
Well that would blow up the scope because I would need to refactor the the API to use something other than `QList`. That is a long-term goal which I wanted to avoid for now because the container is usually very small (less then two dozen elements). I don't think this stopgap is an issue for now.

daschuer commented on src/controllers/bulk/bulksupported.h:
```diff
@@ -1,18 +1,34 @@
 
-#pragma once
+static constexpr std::uint32_t kInvalidInterfaceNumber = 0;
+struct bulk_supported_t {
```
Did you consider to make this a class along with the helper functions as members?

Swiftb0y replied:
Well, which helper functions are you referring to exactly? `is_interesting` should belong to the enumerator IMO.

daschuer replied:
Yes. But I have no special demands here.

Swiftb0y replied:
I don't think moving that is an improvement really. So I'd like to keep as is if thats okay for you.

daschuer replied:
Ok

daschuer reviewed (changes requested):
I have left some more comments.

daschuer commented on src/controllers/bulk/bulkenumerator.cpp:
```diff
@@ -2,60 +2,73 @@
-          m_context(nullptr),
-          m_pConfig(pConfig) {
-    libusb_init(&m_context);
+          m_context(nullptr, &libusb_exit) {
```
The compiler should be smart enough to dispose the function wrapper in the same way as the lambda. Isn't a lambda basically a function without a name but we want a name here? 

Only using the comment works as well if you have performance concerns.

Swiftb0y replied:
well not really performance concerns, more just using the right tool for the job. The solution with the plain function just uses a function pointer (and thus a form of late-dispatch, which we don't need here) while the lambda solution doesn't. [it also results in a size advantage](https://compiler-explorer.com/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:33,endLineNumber:10,positionColumn:33,positionLineNumber:10,selectionStartColumn:33,selectionStartLineNumber:10,startColumn:33,startLineNumber:10),source:'%23include+%3Cmemory%3E%0A%0Avoid+del(int*)%3B%0Aconst+auto+deleterLambda+%3D+%5B%5D(int*+i)+%7Bdel(i)%3B%7D%3B%0A%0Ausing+lambda_ptr+%3D+std::unique_ptr%3Cint,+decltype(deleterLambda)%3E%3B%0Ausing+dynamic_ptr1+%3D+std::unique_ptr%3Cint,+decltype(%26del)%3E%3B%0Ausing+dynamic_ptr2+%3D+std::unique_ptr%3Cint,+void(*)(int*)%3E%3B%0A%0Aint+lambda+%3D+sizeof(lambda_ptr)%3B%0Aint+dptr1+%3D+sizeof(dynamic_ptr1)%3B%0Aint+dptr2+%3D+sizeof(dynamic_ptr2)%3B%0A%0A'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g142,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+14.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+14.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

daschuer replied:
Interesting. The function pointer to the final libusb call must be somewhere though in the lambda case as well.

Swiftb0y replied:
well not really. Its "just" a static call into another function. The linker fills in the address so its the same as any other function call. No indirect call needed in the lambda case (ignoring any implementation details related to cross-library calls).

acolombier commented:
I'll make that a draft PR for now, but feel free to mark it ready when you get the time to pick this work up!
